### ISSUE TITLE
`assert_eq!` fails with `const { .. }` in Rust 2024

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#133081]
This code fails to compile in Rust 2024 when it should pass:

```rust
//@ edition:2024
fn main() {
    assert_eq!(0, const { 0 });
    //~^ ERROR no rules expected keyword `const`
}
```

This will be fixed, I expect, by migrating the standard library to Rust 2024, and when doing that, by using the `:expr` rather than `:expr_2021` macro fragment specifier in `assert_eq` and similar macros.

cc @rust-lang/wg-macros @compiler-errors @ehuss

**Comment by compiler-errors** [comment]
Yeah, this will begin working as soon as the standard library is upgraded to 2024. I don't think there's anything actionable to do here.

**Comment by tgross35** [comment]
Maybe this could go in `tests/crashes` (without the `//~`) so it gets noted once that upgrade happens and it starts passing? Not sure if tests/crashes is only crashes or just a more general xfail.

**Comment by compiler-errors** [comment]
Crashes is for ICEs only

**Comment by fmease** [comment]
cc https://github.com/rust-lang/rust/issues/126457 (corresp. diagnostic issue)

**Comment by traviscross** [comment]
It could go in as a `known-bug` UI test I would expect.

**Comment by ehuss** [comment]
I think this is mostly a duplicate of #126459. No standard library macros with `expr` will work with bare `const{}` until the standard library is updated (regardless of the calling crate's edition). This isn't specific to `assert_eq`. The same goes for any user crates which have exported macros.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
As I said in the meeting:

I consider this the "least bad" option. We ought to explore how to preserve editions for interpolated token trees in macro expansion. However, we ought to do that after this edition has settled down. Seems like a good task for the newly spawned edition subteam. =)

**Comment by ehuss** [comment]
I assume @nikomatsakis's comments were intended for https://github.com/rust-lang/rust/issues/135669.
